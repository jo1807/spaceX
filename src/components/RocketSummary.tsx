import React, { Suspense } from "react";
import styled from "styled-components";

import { CardWrapper } from "./LaunchCard";

interface ILaunchLogo {
  src: string;
}

interface IRocketSummaryProps {
  missionName: string;
  flightNumber: number;
  rocketName: string;
  missionDate: string;
  currentId: number;
  launchId: number;
  launchLogo: string;
  setToggleShowModal: () => void;
}

export const ImageWrapper = styled.div({
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  height: "100%",
  width: "100%",
  margin: "2em",
});

export const LaunchLogo = styled.img<ILaunchLogo>({
  display: "grid",
  width: "10em",
  height: "11em",
  alignItems: "center",
  justifyItems: "center",
  "@media(max-width: 500px)": {
    width: "60%",
    height: "60%",
  },
});

const RocketDetailsButton = styled.button({
  marginTop: "1em",
  backgroundColor: "black",
  color: "white",
  height: "12em",
  width: "10em",
  fontWeight: "bold",
  fontSize: "15px",
  border: "3px solid grey",
  outline: "none",
  ":hover": {
    background: "linear-gradient(white, grey)",
    color: "white",
    cursor: "pointer",
  },
  "@media(max-width: 1024px)": {
    width: "10em",
    height: "5em",
  },
});

const Title = styled.p({
  paddingTop: "2em",
  fontWeight: "bold",
});

const RocketSummary: React.FC<IRocketSummaryProps> = ({
  missionName,
  flightNumber,
  rocketName,
  missionDate,
  currentId,
  launchId,
  launchLogo,
  setToggleShowModal,
}) => {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <CardWrapper
        data-testid={`rocket-summary-${launchId}-id`}
        currentId={currentId}
        launchId={launchId}
      >
        <Title data-testid={`mission-name-${launchId}`}>{missionName}</Title>
        <p data-testid={`flight-number-${launchId}`}>
          Flight Number: {flightNumber}
        </p>
        <p data-testid={`rocket-name-${launchId}`}>Rocket Name: {rocketName}</p>
        <p data-testid={`mission-date-${launchId}`}>{missionDate}</p>
        <RocketDetailsButton
          data-testid={`detailsBtnId-${launchId}`}
          onClick={setToggleShowModal}
        >
          Details
        </RocketDetailsButton>
        <ImageWrapper>
          <LaunchLogo data-testid={`logoId-${launchId}`} src={launchLogo} />
        </ImageWrapper>
      </CardWrapper>
    </Suspense>
  );
};

export default RocketSummary;
