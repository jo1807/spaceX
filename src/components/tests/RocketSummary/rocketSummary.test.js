import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import RocketSummary from "../../RocketSummary";
import { LaunchLogo } from "../../RocketSummary";
import { launches } from "../mockData";
import moment from "moment";

afterEach(cleanup);

it("renders", () => {
  const { asFragment } = render(<RocketSummary />);
  expect(asFragment()).toMatchSnapshot();
});

it("calls correct function on click", () => {
  const onClick = jest.fn();
  const { getByText } = render(<RocketSummary setToggleShowModal={onClick} />);
  fireEvent.click(getByText("Details"));
  expect(onClick).toHaveBeenCalled();
});

it("render logo image on initial render", () => {
  const { queryByTestId } = render(
    <LaunchLogo data-testid="logoId" src={launches[0].images.mission_image} />
  );
  expect(queryByTestId("logoId")).toBeTruthy();
});

it("render prop launches data correctly", () => {
  const missionDate = moment(launches[0].launch_date_local).format(
    "MMMM Do YYYY"
  );
  const { getByTestId } = render(
    <RocketSummary
      missionName={launches[0].mission_name}
      flightNumber={launches[0].flight_number}
      rocketName={launches[0].rocket.rocket_name}
      missionDate={missionDate}
      currentId={launches[0].current_id}
      launchId={launches[0].launch_id}
      launchLogo={launches[0].images.mission_image}
    />
  );
  expect(getByTestId("mission-name").textContent).toBe("DSCOVR");

  expect(getByTestId("flight-number").textContent).toBe("Flight Number: 20");

  expect(getByTestId("rocket-name").textContent).toBe("Rocket Name: Falcon 9");

  expect(getByTestId("mission-date").textContent).toBe("February 11th 2015");
});
