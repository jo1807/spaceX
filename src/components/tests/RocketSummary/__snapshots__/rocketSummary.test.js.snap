// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="css-1n0vo1"
    data-testid="rocket-summary-id"
  >
    <p
      class="css-fd8g0q"
      data-testid="mission-name"
    />
    <p
      data-testid="flight-number"
    >
      Flight Number: 
    </p>
    <p
      data-testid="rocket-name"
    >
      Rocket Name: 
    </p>
    <p
      data-testid="mission-date"
    />
    <button
      class="css-77b2el"
      data-testid="detailsBtnId"
    >
      Details
    </button>
    <div
      class="css-d04zvd"
    >
      <img
        class="css-54d50x"
        data-testid="logoId"
      />
    </div>
  </div>
</DocumentFragment>
`;
